# This is ebay project

In [83]:
import pandas as pd
import numpy as np

In [84]:
autos = pd.read_csv("autos.csv",encoding="Latin-1")

In [85]:
autos.head(5)

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [86]:
autos.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [87]:
autos.head(10)

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5  2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6  2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7  2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8  2016-03-22 16:51:34                                         Seat_Arosa   
9  2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   
5  privat   Angebot  $7,900     test         bus                2006   
6  privat   Angebot    $300     test   limousine                1995   
7  privat   Angebot  $1,990  control   limousine                1998   
8  privat   Angebot    $250     test         NaN                2000   
9  privat   Angebot    $590  control         bus                1997   

     gearbox  powerPS    model   odometer  monthOfRegistration fuelType  \
0    manuell      158   andere  150,000km                    3      lpg   
1  automatik      286      7er  150,000km                    6   benzin   
2    manuell      102     golf   70,000km                    7   benzin   
3  automatik       71   fortwo   70,000km                    6   benzin   
4    manuell        0    focus  150,000km                    7   benzin   
5  automatik      150  voyager  150,000km                    4   diesel   
6    manuell       90     golf  150,000km                    8   benzin   
7    manuell       90     golf  150,000km                   12   diesel   
8    manuell        0    arosa  150,000km                   10      NaN   
9    manuell       90   megane  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   
5    chrysler               NaN  2016-03-21 00:00:00             0   
6  volkswagen               NaN  2016-03-20 00:00:00             0   
7  volkswagen              nein  2016-03-16 00:00:00             0   
8        seat              nein  2016-03-22 00:00:00             0   
9     renault              nein  2016-03-16 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50  
5       22962  2016-04-06 09:45:21  
6       31535  2016-03-23 02:48:59  
7       53474  2016-04-07 03:17:32  
8        7426  2016-03-26 18:18:10  
9       15749  2016-04-06 10:46:35

## Look at column datatypes, column which has nulls, camelcase

In [88]:
print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [89]:
auto_copy = autos.copy()

In [90]:
auto_copy.rename({"yearOfRegistration":"registration_year","monthOfRegistration":"registration_month","notRepairedDamage":"unrepaired_damage","dateCreated":"ad_created"},inplace=True, axis=1)
#auto_copy.rename({"fuelType":"fuel_type","lastSeen":"last_seen","postalCode","postal_code","nrOfPictures":"nr_of_pictures"},inplace=True, axis=1)
auto_copy.rename({"dateCrawled":"date_crawled","vehicleType":"vehicle_type","offerType":"offer_type","fuelType":"fuel_type","lastSeen":"last_seen","postalCode":"postal_code","nrOfPictures":"nr_of_pictures"},inplace=True, axis=1)

In [91]:
auto_copy.columns


Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'powerPS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [92]:
auto_copy.head(5)

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  powerPS   model   odometer  registration_month fuel_type  \
0    manuell      158  andere  150,000km                   3       lpg   
1  automatik      286     7er  150,000km                   6    benzin   
2    manuell      102    golf   70,000km                   7    benzin   
3  automatik       71  fortwo   70,000km                   6    benzin   
4    manuell        0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [93]:
auto_copy.describe(include="all")


date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-14 20:50:02  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox       powerPS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

In [94]:
auto_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
date_crawled          50000 non-null object
name                  50000 non-null object
seller                50000 non-null object
offer_type            50000 non-null object
price                 50000 non-null object
abtest                50000 non-null object
vehicle_type          44905 non-null object
registration_year     50000 non-null int64
gearbox               47320 non-null object
powerPS               50000 non-null int64
model                 47242 non-null object
odometer              50000 non-null object
registration_month    50000 non-null int64
fuel_type             45518 non-null object
brand                 50000 non-null object
unrepaired_damage     40171 non-null object
ad_created            50000 non-null object
nr_of_pictures        50000 non-null int64
postal_code           50000 non-null int64
last_seen             50000 non-null object
dtypes: int64(5)

In [95]:
print(auto_copy.columns)
auto_copy["price"]  = auto_copy["price"].str.replace("$","")
auto_copy["price"]  = auto_copy["price"].str.replace(",","")
auto_copy["price"] = auto_copy["price"].astype(float)

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'powerPS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')


In [96]:
auto_copy["odometer"]  = auto_copy["odometer"].str.replace("km","")
auto_copy["odometer"]  = auto_copy["odometer"].str.replace(",","")
auto_copy["odometer"] = auto_copy["odometer"].astype(float)


In [97]:
auto_copy.rename({"odometer":"odometer_km"},inplace=True, axis=1)

In [98]:
auto_copy["price"].unique().shape

(2357,)

In [99]:
auto_copy["price"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [100]:
auto_copy["odometer_km"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

#### Observation - Note that values are rounded to 1000's, this could be because the drop down options in website have fixed numbers.

### Outliers Detection

In [101]:
auto_copy["price"].value_counts().sort_index(ascending=True).head(20)

0.0     1421
1.0      156
2.0        3
3.0        1
5.0        2
8.0        1
9.0        1
10.0       7
11.0       2
12.0       3
13.0       2
14.0       1
15.0       2
17.0       3
18.0       1
20.0       4
25.0       5
29.0       1
30.0       7
35.0       1
Name: price, dtype: int64

In [102]:
auto_copy["price"].value_counts().sort_index(ascending=False).head(20)

99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
350000.0      1
345000.0      1
299000.0      1
295000.0      1
265000.0      1
259000.0      1
250000.0      1
220000.0      1
198000.0      1
197000.0      1
Name: price, dtype: int64

#### Observation - We can see that there are cars at 0 dollars and car at 1 million dollars. We can remove these outliers to get a more balanced data.

In [103]:
auto_copy = auto_copy[auto_copy["price"].between(1,200000)]
auto_copy["price"].describe()

count     48557.000000
mean       5842.888914
std        8301.871663
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7450.000000
max      198000.000000
Name: price, dtype: float64

### Dropping Columns that have same values throughout, they wont help much

In [104]:
print(auto_copy["nr_of_pictures"].value_counts())
print(auto_copy["seller"].value_counts())
print(auto_copy["offer_type"].value_counts())
auto_copy = auto_copy.drop(["nr_of_pictures", "seller", "offer_type"], axis=1)

0    48557
Name: nr_of_pictures, dtype: int64
privat        48556
gewerblich        1
Name: seller, dtype: int64
Angebot    48557
Name: offer_type, dtype: int64


### Analyse the Years

In [105]:
auto_copy["registration_year"].describe()

count    48557.000000
mean      2004.754968
std         88.651037
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

#### Observation - Note the min column value as 1000, which is too old for a car registration year and max value as 9999, year which has not comer here. These rows could be removed.

In [106]:
ad_cr = auto_copy["ad_created"].value_counts(normalize=True,dropna=False)
ad_cr.sort_index()


2015-06-11 00:00:00    0.000021
2015-08-10 00:00:00    0.000021
2015-09-09 00:00:00    0.000021
2015-11-10 00:00:00    0.000021
2015-12-05 00:00:00    0.000021
                         ...   
2016-04-03 00:00:00    0.038841
2016-04-04 00:00:00    0.036864
2016-04-05 00:00:00    0.011801
2016-04-06 00:00:00    0.003254
2016-04-07 00:00:00    0.001256
Name: ad_created, Length: 76, dtype: float64

### Removing Outliers - Verifying the Reg Year

In [107]:
(~auto_copy["registration_year"].between(1950,2016)).sum() / autos.shape[0]

0.03808

In [108]:
auto_copy = auto_copy[auto_copy["registration_year"].between(1950,2016)]
auto_copy["registration_year"].value_counts(normalize=True).head(15)

2000    0.067648
2005    0.062933
1999    0.062097
2004    0.057938
2003    0.057853
2006    0.057231
2001    0.056481
2002    0.053287
1998    0.050651
2007    0.048807
2008    0.047457
2009    0.044692
1997    0.041819
2011    0.034789
2010    0.034060
Name: registration_year, dtype: float64

#### Observation - Most of the cars were registered in the last 20 years

### Exploring Brands

#### German cars - Volkswagen, BMW, Opel, Mercedes Benz, Audi. USA cars - Ford, France cars - Renault,Peugeot,Fiat, Spain - Seat , Czech - Skoda, Japan - NISSAN,Mazda 

In [109]:
auto_copy["brand"].value_counts(normalize=True)


volkswagen        0.211369
bmw               0.110047
opel              0.107582
mercedes_benz     0.096478
audi              0.086618
ford              0.069878
renault           0.047157
peugeot           0.029859
fiat              0.025658
seat              0.018284
skoda             0.016419
nissan            0.015283
mazda             0.015197
smart             0.014168
citroen           0.014018
toyota            0.012711
hyundai           0.010032
sonstige_autos    0.009624
volvo             0.009153
mini              0.008767
mitsubishi        0.008231
honda             0.007845
kia               0.007074
alfa_romeo        0.006645
porsche           0.006023
suzuki            0.005937
chevrolet         0.005702
chrysler          0.003515
dacia             0.002636
daihatsu          0.002508
jeep              0.002272
subaru            0.002143
land_rover        0.002101
saab              0.001650
jaguar            0.001565
daewoo            0.001500
trabant           0.001372
r

In [110]:
brand_counts = auto_copy["brand"].value_counts(normalize=True)
popular_brands = brand_counts[brand_counts > .03].index
print(popular_brands) 

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford',
       'renault'],
      dtype='object')


#### Finding mean price of common brands

In [111]:
brand_mean_prices = {}

for brand in popular_brands:
    brand_rows = auto_copy[auto_copy["brand"] == brand]
    mean_price = brand_rows["price"].mean()
    brand_mean_prices[brand] = int(mean_price)

brand_mean_prices

{'volkswagen': 5398,
 'bmw': 8236,
 'opel': 2976,
 'mercedes_benz': 8625,
 'audi': 9336,
 'ford': 3737,
 'renault': 2475}

### Exploring Mileage




In [112]:
bmp_series = pd.Series(brand_mean_prices)
pd.DataFrame(bmp_series, columns=["mean_price"])

mean_price
volkswagen           5398
bmw                  8236
opel                 2976
mercedes_benz        8625
audi                 9336
ford                 3737
renault              2475

In [113]:
brand_mean_mileage = {}

for brand in popular_brands:
    brand_only = auto_copy[auto_copy["brand"] == brand]
    mean_mileage = brand_only["odometer_km"].mean()
    brand_mean_mileage[brand] = int(mean_mileage)

mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending=False)
mean_prices = pd.Series(brand_mean_prices).sort_values(ascending=False)

In [114]:
brand_info = pd.DataFrame(mean_mileage,columns=['mean_mileage'])
brand_info
brand_info["mean_price"] = mean_prices
brand_info

mean_mileage  mean_price
bmw                  132644        8236
mercedes_benz        130832        8625
opel                 129336        2976
audi                 129157        9336
volkswagen           128710        5398
renault              128127        2475
ford                 124374        3737

In [115]:
auto_copy.columns


Index(['date_crawled', 'name', 'price', 'abtest', 'vehicle_type',
       'registration_year', 'gearbox', 'powerPS', 'model', 'odometer_km',
       'registration_month', 'fuel_type', 'brand', 'unrepaired_damage',
       'ad_created', 'postal_code', 'last_seen'],
      dtype='object')

In [116]:
auto_copy['postal_code'].value_counts().head(10)

10115    108
65428     98
45888     48
66333     48
44145     46
65933     43
65719     43
48599     42
77933     40
51065     39
Name: postal_code, dtype: int64

#### Observation - 10115 is Berlin. Although, Berlin is thought of walkable city, car remains as a leading way for residents to get around. 65428 is Rüsselsheim, 45888 is Gelsenkirchen, 66333 is Völklingen, 44145 is Dortmund. One can notice in google maps that the public transport is a little less, people tend to use cars as preferred mode of travel.

### Question - How much cheaper are cars with damage than their non-damaged counterparts?

In [119]:
auto_copy['unrepaired_damage'].value_counts()

no_damage_cars = auto_copy[auto_copy['unrepaired_damage'] =='nein']
no_damage_cars_price = no_damage_cars['price'].mean()

damage_cars = auto_copy[auto_copy['unrepaired_damage'] =='ja']
damage_cars_price = damage_cars['price'].mean()

print(no_damage_cars_price)
print(damage_cars_price)


7103.42674669584
2222.3098808998675


#### We can infer from the above the undamaged cars are 2.5 times pricier than the damaged ones.

In [120]:
auto_copy.tail(20)

date_crawled                                               name  \
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_2.0_TDI_DPF_Ambition__NAVI__...   
49988  2016-03-28 19:49:51                                         BMW_330_Ci   
49989  2016-03-11 19:50:37        VW_Polo_zum_Ausschlachten_oder_Wiederaufbau   
49990  2016-03-21 19:54:19         Mercedes_Benz_A_200__BlueEFFICIENCY__Urban   
49991  2016-03-06 15:25:19                                         Kleinwagen   
49992  2016-03-10 19:37:38              Fiat_Grande_Punto_1.4_T_Jet_16V_Sport   
49993  2016-03-15 18:47:35           Audi_A3__1_8l__Silber;_schoenes_Fahrzeug   
49994  2016-03-22 17:36:42  Audi_A6__S6__Avant_4.2_quattro_eventuell_Tausc...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

         price   abtest vehicle_type  registration_year    gearbox  powerPS  \
49979  11000.0     test   kleinwagen               2011    manuell       90   
49980    400.0  control        kombi               1995    manuell      105   
49981   2000.0  control        kombi               1998    manuell      115   
49982   1950.0  control   kleinwagen               2004    manuell        0   
49983    600.0     test   kleinwagen               1999    manuell      101   
49985   1000.0  control          NaN               1995  automatik        0   
49986  15900.0  control    limousine               2010  automatik      218   
49987  21990.0  control    limousine               2013    manuell      150   
49988   9550.0  control        coupe               2001    manuell      231   
49989    150.0     test   kleinwagen               1997    manuell        0   
49990  17500.0     test    limousine               2012    manuell      156   
49991    500.0  control          NaN               2016    manuell        0   
49992   4800.0  control   kleinwagen               2009    manuell      120   
49993   1650.0  control   kleinwagen               1997    manuell        0   
49994   5000.0  control        kombi               2001  automatik      299   
49995  24900.0  control    limousine               2011  automatik      239   
49996   1980.0  control       cabrio               1996    manuell       75   
49997  13200.0     test       cabrio               2014  automatik       69   
49998  22900.0  control        kombi               2013    manuell      150   
49999   1250.0  control    limousine               1996    manuell      101   

          model  odometer_km  registration_month fuel_type          brand  \
49979      polo      70000.0                  11    diesel     volkswagen   
49980    escort     125000.0                   3    benzin           ford   
49981     astra     150000.0                  12    benzin           opel   
49982     fabia      90000.0                   7    benzin          skoda   
49983     focus     150000.0                   4    benzin           ford   
49985       NaN     150000.0                   0    benzin     volkswagen   
49986      300c     125000.0                  11    diesel       chrysler   
49987        a3      50000.0        

### Find the most common brand/model combinations

In [121]:
for brand in popular_brands:
    brand_rows = auto_copy[auto_copy["brand"] == brand]
    print("Brand Name: ", brand)
    print(brand_rows["model"].value_counts().head(3)) 

Brand Name:  volkswagen
golf      3707
polo      1609
passat    1349
Name: model, dtype: int64
Brand Name:  bmw
3er    2615
5er    1132
1er     521
Name: model, dtype: int64
Brand Name:  opel
corsa     1591
astra     1348
vectra     544
Name: model, dtype: int64
Brand Name:  mercedes_benz
c_klasse    1136
e_klasse     958
a_klasse     539
Name: model, dtype: int64
Brand Name:  audi
a4    1231
a3     825
a6     797
Name: model, dtype: int64
Brand Name:  ford
focus     762
fiesta    722
mondeo    479
Name: model, dtype: int64
Brand Name:  renault
twingo    615
clio      473
megane    335
Name: model, dtype: int64


### Categorical columns have german words, which need to be converted into english.

In [125]:
categorical_columns = ['name','vehicle_type','fuel_type','unrepaired_damage','gearbox','model']

auto_copy[categorical_columns] = auto_copy[categorical_columns].replace(regex='Noch', value='still').replace(regex='gut', value='well').replace(regex='erhaltenen', value='preserved').replace(regex='inkl', value='incl').replace(regex='Sitzer', value='Seater').replace(regex='gruen', value='green').replace(regex='angemeldet', value='registered').replace(regex='oder', value='or').replace(regex='Silber', value='silver').replace(regex='Leder', value='leather').replace(regex='reihe', value='series').replace(regex='Unfall', value='accident').replace(regex='und', value='and').replace(regex='Verkaufen', value='sell').replace(regex='Klimaanlage', value='Air Conditioner').replace(regex='umbau', value='modification').replace(regex='kombi', value='combi').replace(regex='kleinwagen', value='small car').replace(regex='sehr', value='very').replace(regex='manuell', value='manual').replace(regex='ja', value='yes').replace(regex='voll', value='full').replace(regex='fahrzeug', value='vehicle').replace(regex='gepflegt', value='well looked after')
auto_copy[categorical_columns] = auto_copy[categorical_columns].replace(regex='Motorschaden', value='Engine Damage').replace(regex='Anhaengerkupplung', value='Trailer Hitch').replace(regex='elektro', value='electric').replace(regex='klasse', value='class').replace(regex='neu', value='new').replace(regex='nein', value='no').replace(regex='andere', value='others').replace(regex='ist', value='is').replace(regex='benzin', value='petrol').replace(regex='automatik', value='automatic').replace(regex='mit', value='with').replace(regex='fahrzeug', value='vehicle').replace(regex='cabrio', value='convertible')

In [127]:
auto_copy[categorical_columns][:40]

name vehicle_type fuel_type  \
0                    Peugeot_807_160_NAVTECH_ON_BOARD          bus       lpg   
1          BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik    limousine    petrol   
2                          Volkswagen_Golf_1.6_United    limousine    petrol   
3   Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...    small car    petrol   
4   Ford_Focus_1_6_Benzin_TÜV_new_is_very_well loo...        combi    petrol   
5   Chrysler_Grand_Voyager_2.8_CRD_Aut.Liwithed_St...          bus    diesel   
6   VW_Golf_III_GT_Special_Electronic_Green_Metall...    limousine    petrol   
7                                Golf_IV_1.9_TDI_90PS    limousine    diesel   
8                                          Seat_Arosa          NaN       NaN   
9       Renault_Megane_Scenic_1.6e_RT_Air Conditioner          bus    petrol   
11                        Mercedes_A140_Engine Damage          NaN    petrol   
12  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...    small car    petrol   
13                                 Audi_A3_1.6_tuning    limousine    petrol   
14  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...    small car    petrol   
15        Corvette_C3_Coupe_T_Top_Crossfire_Injection        coupe    petrol   
16                                Opel_Vectra_B_Kombi        combi    petrol   
17                          Volkswagen_Scirocco_2_G60        coupe    petrol   
18                    sell_mein_bmw_e36_320_i_touring          bus    petrol   
19       mazda_tribute_2.0_with_gas_and_tuev_new_2018          suv       lpg   
20  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...        combi    diesel   
21                      Porsche_911_Carrera_4S_Cabrio  convertible    petrol   
22                               MINI_Cooper_S_Cabrio  convertible    petrol   
23   Peugeot_Boxer_2_2_HDi_120_Ps_9_Seater_incl_Klima          bus    diesel   
24                         BMW_535i_xDrive_Sport_Aut.    limousine    petrol   
25  Ford_escort_combi_an_bastler_with_ghia_ausstat...        combi    petrol   
26                                Volkswagen_Polo_Fox    small car    petrol   
28                                      MINI_Cooper_D    small car    diesel   
29  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...        combi    diesel   
30                        Peugeot_206_accidentvehicle    small car    petrol   
31                         still_well_preserved_C_320        combi    petrol   
32                              Corsa_with_TÜV_5.2016    small car    petrol   
33  Audi_A3_2.0_TDI_Sportback_DPF_Kupplung_defekt_...        combi    diesel   
34                            Bmw_530d__navi_tuv03/18        combi    diesel   
35                                        Ford_Mondeo        combi    petrol   
36          Ford_Mondeo_2.0_Turnier_TDCi_DPF_Titanium        combi    diesel   
37                 Seat_Altea_XL_2.0_TDI_CR_DPF_Style          bus    diesel   
38                        Volkswagen_Golf_1.4_Special    limousine    petrol   
39                 Mercedes_C_180_Kombi__zu_verkaufen        combi    petrol   
40  VW_Golf__4_Cabrio_2.0_GTI_16V___leather_MFA_Al...  convertible    petrol   
41  Passat_3b_1.9_TDI_Highline__regisered_with_tuv...        combi       NaN   

   unrepaired_damage    gearbox     model  
0                 no     manual    others  
1                 no  automatic       7er  
2                 no     manual      golf  
3                 no  automatic    fortwo  
4                 no     manual     focus  
5                NaN  automatic   voyager  
6                NaN     manual      golf  
7                 no     manual      golf  
8                 no     manual     arosa  
9                 no     manual    megane  
11               NaN        NaN       NaN  
12                no  automatic    fortwo  
13                no     manual        a3  
14               NaN     manual      clio  
15                no  automatic       NaN  
16                no     manual    vectra  
17                no

### Split the odometer_km into groups, and use aggregation to see if average prices follows any patterns based on the mileage.
    

In [146]:
auto_copy["odometer_km"].value_counts()


150000.0    30083
125000.0     4857
100000.0     2055
90000.0      1672
80000.0      1375
70000.0      1187
60000.0      1126
50000.0       992
40000.0       796
5000.0        771
30000.0       759
20000.0       740
10000.0       240
Name: odometer_km, dtype: int64

In [147]:
ac = auto_copy
bins = [0, 30000, 60000, 80000, 100000,125000, 150000]
names = ['<30k', '30k-60k', '60k-80k', '80k-100k', '100k-125k','125k-150k']

ac['odometer_km_range'] = pd.cut(ac['odometer_km'], bins, labels=names)

In [141]:
ac['odometer_km_range'].value_counts()

125k-150k    30083
100k-125k     4857
80k-100k      3727
30k-60k       2914
60k-80k       2562
<30k          2510
Name: odometer_km_range, dtype: int64

In [152]:
ac[(ac['odometer_km'] >= 30000) & (ac['odometer_km']  < 60000)].mean()

price                 15090.836278
registration_year      2009.001570
powerPS                 128.509619
odometer_km           40914.801728
registration_month        6.019238
postal_code           52778.246957
dtype: float64

In [154]:
ac[(ac['odometer_km'] >= 125000) & (ac['odometer_km']  < 150000)].mean()

price                   6214.022030
registration_year       2003.840642
powerPS                  116.596665
odometer_km           125000.000000
registration_month         5.974264
postal_code            51950.810171
dtype: float64

In [155]:
ac[(ac['odometer_km'] >= 80000) & (ac['odometer_km']  < 100000)].mean()

price                  9034.021989
registration_year      2005.332786
powerPS                 124.231375
odometer_km           85487.364621
registration_month        6.059731
postal_code           51910.091237
dtype: float64

In [156]:
ac[(ac['odometer_km'] >= 60000) & (ac['odometer_km']  < 80000)].mean()

price                 11628.383485
registration_year      2006.594898
powerPS                 129.191094
odometer_km           65131.863381
registration_month        6.163424
postal_code           52616.250757
dtype: float64

#### Observation - We can derive that price is inversely proportional to odometer ratings.